src/preamble.js

```diff
@@ -134,11 +134,55 @@ var HEAP,
 var HEAP_DATA_VIEW;
 #endif
 
+function fixPointer (ptr) {
+  if (typeof ptr === 'number' && ptr < 0) {
+    return ptr >>> 0;
+  }
+  if (typeof ptr === 'bigint') {
+    return Number(ptr);
+  }
+  return ptr;
+}
+
+const getHeapBlock =  (type, offset, length) => {
+  const maxSize = Math.min(wasmMemory.buffer.byteLength, 4 * 1024 * 1024 * 1024 - 8)
+
+  const heap = wasmMemory.buffer
+  switch (type) {
+    case 'i1':
+    case 'i8':
+      return new Int8Array(heap, offset, length || maxSize / Int8Array.BYTES_PER_ELEMENT);
+    case 'u1':
+    case 'u8':
+      return new Uint8Array(heap, offset, length || maxSize / Uint8Array.BYTES_PER_ELEMENT);
+    case 'i16':
+      return new Int16Array(heap, offset, length || maxSize / Int16Array.BYTES_PER_ELEMENT);
+    case 'u16':
+      return new Uint16Array(heap, offset, length || maxSize / Uint16Array.BYTES_PER_ELEMENT);
+    case 'i32':
+      return new Int32Array(heap, offset, length || maxSize / Int32Array.BYTES_PER_ELEMENT);
+    case 'u32':
+      return new Uint32Array(heap, offset, length || maxSize / Uint32Array.BYTES_PER_ELEMENT);
+    case 'f32':
+      return new Float32Array(heap, offset, length || maxSize / Float32Array.BYTES_PER_ELEMENT);
+    case 'f64':
+      return new Float64Array(heap, offset, length || maxSize / Float64Array.BYTES_PER_ELEMENT);
+    case 'i64':
+      return new BigInt64Array(heap, offset, length || maxSize / BigInt64Array.BYTES_PER_ELEMENT);
+    case '*':
+    case 'u64':
+      return new BigUint64Array(heap, offset, length || maxSize / BigUint64Array.BYTES_PER_ELEMENT);
+    default:
+      throw new Error('Invalid type');
+  }
+}
+
 function updateMemoryViews() {
   var b = wasmMemory.buffer;
 #if SUPPORT_BIG_ENDIAN
   Module['HEAP_DATA_VIEW'] = HEAP_DATA_VIEW = new DataView(b);
 #endif
+#if !WASM_BIGINT
```

Comment:
> I think this should probably be "if !MEMORY64` no ?

yeah

i've just tested the other HEAPs in chrome with initial memory=8gb and it throws `Invalid typed array length: undefined`, so looks like it's all or nothing. seems it's tied with buffer length, not array length itself
![image](https://github.com/emscripten-core/emscripten/assets/7466130/647ab81c-d9c3-4f43-b13d-203f2ce57811)

Reply:
My understanding from the v8 engine folks that the limit is on the number elements, no the size of the underlying buffer.  If that isn't true we should probably update the upstream bug: https://bugs.chromium.org/p/v8/issues/detail?id=4153

Reply:
You're right. The issue is even seen on my screenshot, i was creating Uint8Array for HEAPU8

So if memory size is less than 8gb only HEAP(U)8 should be proxied and HEAP(U)16 otherwise. Will add some tests and conditional proxy creation